File: library/src/main/java/com/dylanc/retrofit/helper/interceptor/DebugInterceptor.java
package com.dylanc.retrofit.helper.interceptor;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.RawRes;
import okhttp3.*;

import java.io.*;

public class DebugInterceptor implements Interceptor {
  private final Context mContext;
  private final String mDebugUrl;
  private final int mDebugRawId;

  public DebugInterceptor(Context context, String debugUrl, int debugRawId) {
    mContext = context;
    mDebugUrl = debugUrl;
    mDebugRawId = debugRawId;
  }

  @NonNull
  @Override
  public Response intercept(@NonNull Chain chain) throws IOException {
    final String url = chain.request().url().toString();
    if (url.contains(mDebugUrl)) {
      return debugResponse(chain, mDebugRawId);
    }
    return chain.proceed(chain.request());
  }

  private Response debugResponse(Chain chain, @RawRes int rawId) {
    final String json = getRawFile(mContext, rawId);
    return createResponse(chain, json);
  }

  private Response createResponse(Chain chain, String json) {
    return new Response.Builder()
        .code(200)
        .addHeader("Content-Type", "application/json")
        .body(ResponseBody.create(MediaType.parse("application/json"), json))
        .message("OK")
        .request(chain.request())
        .protocol(Protocol.HTTP_1_1)
        .build();
  }

  private static String getRawFile(Context context, int id) {
    final InputStream is = context.getResources().openRawResource(id);
    final BufferedInputStream bis = new BufferedInputStream(is);
    final InputStreamReader isr = new InputStreamReader(bis);
    final BufferedReader br = new BufferedReader(isr);
    final StringBuilder stringBuilder = new StringBuilder();
    String str;
    try {
      while ((str = br.readLine()) != null) {
        stringBuilder.append(str);
      }
    } catch (IOException e) {
      e.printStackTrace();
    } finally {
      try {
        br.close();
        isr.close();
        bis.close();
        is.close();
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
    return stringBuilder.toString();
  }
}

File: library/src/main/java/com/dylanc/retrofit/helper/interceptor/DebugInterceptor.kt
package com.dylanc.retrofit.helper.interceptor

import android.content.Context
import android.support.annotation.RawRes
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Protocol
import okhttp3.Response
import okhttp3.ResponseBody.Companion.toResponseBody
import java.io.IOException

/**
 * @author Dylan Cai
 * @since 2019/7/13
 */
class DebugInterceptor(private val context: Context, private val debugUrl: String, private val debugRawId: Int) :
  Interceptor {

  @Throws(IOException::class)
  override fun intercept(chain: Interceptor.Chain): Response {
    val url = chain.request().url.toString()
    return if (url.contains(debugUrl)) {
      getDebugResponse(chain, debugRawId)
    } else {
      chain.proceed(chain.request())
    }
  }

  private fun getDebugResponse(chain: Interceptor.Chain, @RawRes rawId: Int): Response {
    val inputStream = context.resources.openRawResource(rawId)
    val json = inputStream.bufferedReader().use { it.readText() }
    return Response.Builder()
      .code(200)
      .addHeader("Content-Type", "application/json")
      .body(json.toResponseBody("application/json".toMediaTypeOrNull()))
      .message("OK")
      .request(chain.request())
      .protocol(Protocol.HTTP_1_1)
      .build()
  }
}